hydradx/tests/test_centralized_market.py

```diff
@@ -0,0 +1,246 @@
+    trade_fee=st.floats(min_value=0, max_value=0.1),
+    order_book=order_book_strategy()
+)
+def test_buy_quote(quantity: float, trade_fee: float, order_book: OrderBook):
+    initial_state = GlobalState(
+        pools={
+            'Kraken': CentralizedMarket(
+                order_book={
+                    ('ETH', 'DAI'): order_book
+                }, trade_fee=trade_fee
+            )
+        },
+        agents={'agent': initial_agent}
+    )
+    tkn_buy = 'DAI'
+    tkn_sell = 'ETH'
+    buy_state = initial_state.copy()
+    buy_state.execute_swap(
+        pool_id='Kraken',
+        agent_id='agent',
+        tkn_sell=tkn_sell,
+        tkn_buy=tkn_buy,
+        buy_quantity=quantity
+    )
+
+    value_sold = sum(
+        [quantity * price for (price, quantity) in initial_state.pools['Kraken'].order_book[(tkn_sell, tkn_buy)].asks]
+    ) - sum(
+        [quantity * price for (price, quantity) in buy_state.pools['Kraken'].order_book[(tkn_sell, tkn_buy)].asks]
+    )
+
+    quantity_bought = buy_state.agents['agent'].holdings[tkn_buy] - initial_state.agents['agent'].holdings[tkn_buy]
+
+    if value_sold != pytest.approx(quantity_bought):
+        raise AssertionError('Central market buy trade failed to execute correctly.')
+
+    quantity_sold = initial_state.agents['agent'].holdings[tkn_sell] - buy_state.agents['agent'].holdings[tkn_sell]
+
+    value_bought = sum(
+        [quantity for (price, quantity) in initial_state.pools['Kraken'].order_book[(tkn_sell, tkn_buy)].asks]
+    ) - sum(
+        [quantity for (price, quantity) in buy_state.pools['Kraken'].order_book[(tkn_sell, tkn_buy)].asks]
+    )
+
+    if value_bought / (1 - initial_state.pools['Kraken'].trade_fee) != pytest.approx(quantity_sold):
+        raise AssertionError('Fee was not applied correctly.')
+
+
+@given(
+    quantity=st.floats(min_value=0.1, max_value=1000),
+    order_book=order_book_strategy()
+)
+def test_buy_base(quantity: float, order_book: OrderBook):
+    initial_state = GlobalState(
+        pools={
+            'Kraken': CentralizedMarket(
+                order_book={
+                    ('ETH', 'DAI'): order_book
+                },
+            )
+        },
+        agents={'agent': initial_agent}
+    )
+    tkn_buy = 'ETH'
+    tkn_sell = 'DAI'
+    sell_state = initial_state.copy()
+    sell_state.execute_swap(
+        agent_id='agent',
+        pool_id='Kraken',
+        tkn_sell=tkn_sell,
+        tkn_buy=tkn_buy,
+        buy_quantity=quantity
+    )
+
+    quantity_bought = sell_state.agents['agent'].holdings[tkn_buy] - initial_state.agents['agent'].holdings[tkn_buy]
+
+    value_sold = sum(
+        [quantity for (price, quantity) in initial_state.pools['Kraken'].order_book[(tkn_buy, tkn_sell)].bids]
+    ) - sum(
+        [quantity for (price, quantity) in sell_state.pools['Kraken'].order_book[(tkn_buy, tkn_sell)].bids]
+    )
+
+    if value_sold != pytest.approx(quantity_bought):
```

Comment:
Is there a reason that we are testing with pytest.approx? I think this should be exact? There's only additions/subtractions involved, right?

Reply:
There is multiplication by the bid/ask price and trade fee, so there does end up being a rounding error.